In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

df = pd.read_excel("Online retail.xlsx", header=None)
df.head()

0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey    avocado         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1               NaN   NaN             NaN           NaN           NaN   
2               NaN   NaN             NaN           NaN           NaN   
3               NaN   NaN             NaN           NaN           NaN   
4               NaN   NaN             NaN           NaN           NaN   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1             NaN        NaN    NaN    NaN            NaN     NaN   
2             NaN        NaN    NaN    NaN            NaN     NaN   
3             NaN        NaN    NaN    NaN            NaN     NaN   
4             NaN        NaN    NaN    NaN            NaN     NaN   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN  
2                NaN              NaN      NaN        NaN  
3                NaN              NaN      NaN        NaN  
4                NaN              NaN      NaN        NaN

In [2]:
transactions = []

for i in range(0, 7501):
    transactions.append([str(df.values[i, j]) for j in range(20)])
len(transactions)

7501

In [3]:
transactions[0]

['shrimp',
 'almonds',
 'avocado',
 'vegetables mix',
 'green grapes',
 'whole weat flour',
 'yams',
 'cottage cheese',
 'energy drink',
 'tomato juice',
 'low fat yogurt',
 'green tea',
 'honey',
 'salad',
 'mineral water',
 'salmon',
 'antioxydant juice',
 'frozen smoothie',
 'spinach',
 'olive oil']

In [4]:
transactions[7500]

['eggs',
 'frozen smoothie',
 'yogurt cake',
 'low fat yogurt',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan']

In [5]:
from apyori import apriori
rules = apriori(transactions, min_support = 0.003, min_confidence = 0.2, min_lift = 3,
                min_length = 2, max_length = 2)
results = list(rules)
len(results)

9

In [6]:
results

[RelationRecord(items=frozenset({'chicken', 'light cream'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]),
 RelationRecord(items=frozenset({'escalope', 'mushroom cream sauce'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)]),
 RelationRecord(items=frozenset({'escalope', 'pasta'}), support=0.005865884548726837, ordered_statistics=[OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.3728813559322034, lift=4.700811850163794)]),
 RelationRecord(items=frozenset({'fromage blanc', 'honey'}), support=0.003332888948140248, ordered_statistics=[OrderedStatistic(items_base=frozenset({'fromage blanc'}), items_add=frozenset({'honey'}), confidence=0

In [7]:
print('Support:', results[0][1]) # Support
print('Base item:', results[0][2][0][0]) # Base item
print('Add item1:', results[0][2][0][1]) # Add item
print('Confidence:', results[0][2][0][2]) # Confidence
print('Lift:', results[0][2][0][3]) # Lift

Support: 0.004532728969470737
Base item: frozenset({'light cream'})
Add item1: frozenset({'chicken'})
Confidence: 0.29059829059829057
Lift: 4.84395061728395


In [8]:
a = []
b = []
c = []
d = []
e = []

for i in range(0, 9):
    c.append(results[i][1]) # Support
    a.append(results[i][2][0][0]) # Base item
    b.append(results[i][2][0][1]) # Add item
    d.append(results[i][2][0][2]) # Confidence
    e.append(results[i][2][0][3])

d1 = pd.DataFrame(a)
d2 = pd.DataFrame(b)
d3 = pd.DataFrame(c)
d4 = pd.DataFrame(d)
d5 = pd.DataFrame(e)

new_data = pd.concat([d1, d2, d3, d4, d5], axis = 1)
new_data.columns = ['Base item', 'Add item', 'Support', 'Confidence', 'Lift']
new_data

Base item     Add item   Support  Confidence      Lift
0           light cream      chicken  0.004533    0.290598  4.843951
1  mushroom cream sauce     escalope  0.005733    0.300699  3.790833
2                 pasta     escalope  0.005866    0.372881  4.700812
3         fromage blanc        honey  0.003333    0.245098  5.164271
4         herb & pepper  ground beef  0.015998    0.323450  3.291994
5          tomato sauce  ground beef  0.005333    0.377358  3.840659
6           light cream    olive oil  0.003200    0.205128  3.114710
7     whole wheat pasta    olive oil  0.007999    0.271493  4.122410
8                 pasta       shrimp  0.005066    0.322034  4.506672

From the above analysis of the transactions, the confidence of 'Ground beef' for buying 'Tomato Sauce' is 0.377.

Which means the probability of buying 'Ground beef' is more likely when 'Tomato sauce' is purchased.

And also the lift ratio of 'Fromage blanc' and 'honey' is more compared to all other transactions.